### PR TITLE
Explorer - Fix Retaining of filters on switching between Trace and Span view

### PR DESCRIPTION
**Pull request by jaywalker21**
## Description
***Fixes https://github.com/hypertrace/hypertrace-ui/issues/1250***
- Made the `filter-bar` component controlled in nature by providing a mechanism for two-way binding.
- Added logic in `explorer` component to update the filter values on toggling between `span` and `trace` views.

### Testing
- Carried out local testing and seems to be working well so far.

### CODE REVIEW

File: projects/observability/src/pages/explorer/explorer.component.ts [hunk 1]
```diff
@@ -194,11 +198,51 @@ export class ExplorerComponent {
...
+        type: applicableType,
+        onlySupportsAggregation: onlySupportsAggregation,
+        onlySupportsGrouping: onlySupportsGrouping
+      };
+    });
+  }
+
   public onContextUpdated(contextWrapper: ExplorerContextScope): void {
     this.attributes$ = this.metadataService.getFilterAttributes(contextWrapper.dashboardContext);
+    const listener = this.attributes$.subscribe(attributes => {
```
Comment (jaywalker21) [inline projects/observability/src/pages/explorer/explorer.component.ts:218]: Will try to move this block outside of this method in an independent structure.

File: projects/observability/src/pages/explorer/explorer.component.ts [hunk 2]
```diff
@@ -327,6 +371,12 @@ interface ExplorerContextScope {
   scopeQueryParam: ScopeQueryParam;
 }
 
+type contextMap = {
```
Comment (aaron-steinfeld) [inline projects/observability/src/pages/explorer/explorer.component.ts:374]: PascalCase for type names (also, we generally use interfaces if constructing a brand new type that's not projected from another.

Naming here is also an issue. I think context is an overloaded term, and even knowing the goal of this PR it took me reading the object using this to understand what this was. Maybe something like
``` ts
type AttributeTranslationDictionary = { [key: string]: string; }

const spanToApiTraceAttributeTranslationDictionary = {...};
const apiTraceToSpanAttributeTranslationDictionary = {...};
```

File: projects/observability/src/pages/explorer/explorer.component.ts (hunk 1 above)
Comment (aaron-steinfeld) [inline projects/observability/src/pages/explorer/explorer.component.ts:218]: We shouldn't use a subscription here. Make the filters an observable, and subscribe via async pipe in the template.

File: projects/observability/src/pages/explorer/explorer.component.ts [hunk 3]
```diff
@@ -194,11 +198,51 @@ export class ExplorerComponent {
...
+  }
+
   public onContextUpdated(contextWrapper: ExplorerContextScope): void {
     this.attributes$ = this.metadataService.getFilterAttributes(contextWrapper.dashboardContext);
+    const listener = this.attributes$.subscribe(attributes => {
+      const lastTab = this.contextChangeSubject.getValue();
+      const newFilters = this.filters.map(eachFilter => {
+        // If the given filter has a different name for the selected tab, update the filter value
+        if (eachFilter.field in contextMapObject[lastTab]) {
+          const newFilter = this.filterChipService.autocompleteFilters(
```
Comment (aaron-steinfeld) [inline projects/observability/src/pages/explorer/explorer.component.ts:223]: We should be able to do this without the chip service, that's an impl detail from the filter bar. Can we use the filter builder service (I hate how complex we made this!) instead just read the details out of the old one and write them into the new one. May need to lookup the new attribute, too - but we shouldn't have to create it, it's already one of the ones in the array coming in.

File: projects/components/src/filtering/filter-bar/filter-bar.component.ts [hunk 4]
```diff
@@ -175,10 +193,10 @@ export class FilterBarComponent implements OnChanges, OnInit, OnDestroy {
   }
 
   private updateFilter(oldFilter: Filter, newFilter: Filter): void {
-    this.onFiltersChanged(this.filterBarService.updateFilter(this.internalFiltersSubject$.value, oldFilter, newFilter));
+    this.onFiltersChanged(this.filterBarService.updateFilter(this.filters || [], oldFilter, newFilter));
```
Comment (aaron-steinfeld) [inline projects/components/src/filtering/filter-bar/filter-bar.component.ts:196]: When do we use the subject's value and when do we use this.filters? We seem to use different ways in different places which makes it very confusion to understand as a reader.

**Pull request closed without merging**